union: make {ep}lno policy universal across all remote types

ferferga opened the pull request:
#### What is the purpose of this change?

The `{ep}lno` policy of rclone was unsupported by [a lot of backends](https://github.com/rclone/rclone/issues/8492#issuecomment-2791987004). This PR adds support to all of them by falling back to normal listing instead of object counting (informing the user that the operation will be slower).

The log listed in the issue below becomes:

<details>

```
2025/04/10 17:04:59 DEBUG : Setting --user-agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36 Edg/134.0.0.0" from environment variable RCLONE_USER_AGENT="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36 Edg/134.0.0.0"
2025/04/10 17:04:59 DEBUG : rclone: Version "v1.70.0-beta.8712.7a09de9f2.better-lno-policy" starting with parameters ["./rclone" "copy" "-vv" "rclone.conf" "Union:"]
2025/04/10 17:04:59 DEBUG : Creating backend with remote "rclone.conf"
2025/04/10 17:04:59 DEBUG : Using config file from "/home/ferferga/.config/rclone/rclone.conf"
2025/04/10 17:04:59 DEBUG : fs cache: renaming child cache item "rclone.conf" to be canonical for parent "/home/ferferga/.config/rclone"
2025/04/10 17:04:59 DEBUG : Creating backend with remote "Union:"
2025/04/10 17:04:59 DEBUG : Creating backend with remote "Crypt:Backups/1"
2025/04/10 17:04:59 DEBUG : Creating backend with remote "Crypt:Backups/3"
2025/04/10 17:04:59 DEBUG : Creating backend with remote "Crypt:Backups/2"
2025/04/10 17:04:59 DEBUG : Creating backend with remote "OpenDrive:[REDACTED]"
2025/04/10 17:04:59 DEBUG : Creating backend with remote "OpenDrive:[REDACTED]"
2025/04/10 17:04:59 DEBUG : Creating backend with remote "OpenDrive:[REDACTED]"
2025/04/10 17:05:00 DEBUG : Starting OpenDrive session with ID: [REDACTED]
2025/04/10 17:05:00 DEBUG : Starting OpenDrive session with ID: [REDACTED]
2025/04/10 17:05:00 DEBUG : Starting OpenDrive session with ID: [REDACTED]
2025/04/10 17:05:01 DEBUG : union root '': actionPolicy = *policy.Lno, createPolicy = *policy.Lno, searchPolicy = *policy.FF
2025/04/10 17:05:02 DEBUG : rclone.conf: Need to transfer - File not found at Destination
2025/04/10 17:05:02 NOTICE: Number of Objects is not supported for upstream Crypt, falling back to listing (this may be slower)...
2025/04/10 17:05:06 DEBUG : Counted 9233 objects for upstream Crypt by listing
2025/04/10 17:05:06 NOTICE: Number of Objects is not supported for upstream Crypt, falling back to listing (this may be slower)...
2025/04/10 17:05:06 DEBUG : Counted 27 objects for upstream Crypt by listing
2025/04/10 17:05:07 DEBUG : [REDACTED]: Uploading chunk 0, size=2542536, remain=0
2025/04/10 17:05:09 DEBUG : rclone.conf: md5 = 400d21863fca7f183fc80b95a5dd0a14 OK
2025/04/10 17:05:09 INFO  : rclone.conf: Copied (new)
```

</details>

#### Was the change discussed in an issue or in the forum before?

Fixes #8492 

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate — *(this didn't have any tests and I don't have experience with Go to feel safe to do good ones, but I will happily take feedback)*
- [ ] I have added documentation for the changes if appropriate — *(will do once code review is over)*
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

ferferga commented on backend/union/policy/eplno.go:
```diff
@@ -91,7 +100,7 @@ func (p *EpLno) CreateEntries(entries ...upstream.Entry) ([]upstream.Entry, erro
 		return nil, err
 	}
-	e, err := p.lnoEntries(entries)
+	e, err := p.lnoEntries(context.Background(), entries)
```
Not sure if this way of instantiating a new context is appropiate in this context, this is my first time doing a bit more of Go and a background context seemed to me the most sensible for this use case, but I actually have no idea 😅.

ncw replied:
You either use `context.Background()` if this is really a background operation, or `context.TODO()` to say you'd really like to have a context here but you don't have one, or you go context plumbing and plumb in ctx into this function and all its callers etc.

If you look at the calls to the About backend method in the upstream.go file you'll see it is called from the background so assuming this code gets moved into upstream.go then it should be `context.Background()`

ncw commented on backend/union/policy/eplno.go:
```diff
@@ -18,15 +18,28 @@ type EpLno struct {
+		uName := upstream.Name()
+		fs.LogPrintf(fs.LogLevelNotice, nil,
+			"Number of Objects is not supported for upstream %s, falling back to listing (this may be slower)...", uName)
+		listing, err := upstream.List(ctx, "")
```
This should use this function

https://github.com/rclone/rclone/blob/e0c99d620340f974255820882ab2802c6694a9ab/fs/operations/operations.go#L1009-L1012

Otherwise you've just listed the root directory which I don't think is what you want to do

ncw commented on backend/union/policy/eplno.go:
```diff
@@ -18,15 +18,28 @@ type EpLno struct {
+			"Number of Objects is not supported for upstream %s, falling back to listing (this may be slower)...", uName)
+		listing, err := upstream.List(ctx, "")
+		if err != nil {
+			fs.LogPrintf(fs.LogLevelError, nil, "Listing failed, treating as 0: %v", err)
```
fs.Errorf is the same as fs.LogPrintf(fs.LogLevelError - same for the other debug levels

ncw reviewed (commented):
Interesting idea and glad to welcome you to the world of Go programming :-)

I put some comments about the code inline.

Several things occur to me about this

1. It is going to be really inefficient!
2. We have a feature like this in the VFS already

      --vfs-used-is-size rclone size           Use the rclone size algorithm for Used size

Which would be worth studying

3. This potentially is useful for other policies also - the size based ones, because not all backends support About read the size used (eg s3)

ferferga commented:
@ncw 
1. I know — see my stance on it [here](https://github.com/rclone/rclone/issues/8492#issuecomment-2794277896)
2. Is the VFS available for backends? Can you clarify further how you think that would work? I thought VFS was only available at the user level through CLI flags and specifically for mount (and the approach needs to be usable for all rclone commands).

ferferga commented on backend/union/policy/eplno.go:
```diff
@@ -18,15 +18,28 @@ type EpLno struct {
+func (p *EpLno) getUpstreamObjectCount(ctx context.Context, upstream *upstream.Fs) int64 {
+	numObj, err := upstream.GetNumObjects()
+	if err != nil {
+		uName := upstream.Name()
```
This was exactly the first place I thought about implementing this! But given what would be returned by this function are not the "objects" per-se of the backend, I thought it was more appropiate implementing it in the policy code itself. Will apply the refactor in my next commit 👍.

ncw commented:
@ferferga 
> 2. Is the VFS available for backends? Can you clarify further how you think that would work? I thought VFS was only available at the user level through CLI flags and specifically for mount (and the approach needs to be usable for all rclone commands).

Yes, you can't use the VFS code directly, but I thought it might be interesting for you to look at.

ferferga commented:
@ncw With this new commit, everything is properly hooked with the already implemented caching mechanism. Also:
* I moved the NOTICE to WARNING level so the user could take an action if slower behaviour is undesired and it's best for him to not use a policy at all
* Simplified error handling by check for `err` only once in the function's.
About the VFS: I checked [the implementation](https://github.com/rclone/rclone/blob/master/vfs/vfs.go#L623-631), but under the hood is already using `Count` (because it's what `rclone size` does). We also don't care about the size, but only the number of objects.

ferferga commented on docs/content/union.md:
```diff
@@ -155,17 +155,21 @@ as necessary.
+| lno, eplno | Objects, Used¹     |
+
+¹ Objects are only available for S3-based upstreams. For better compatibility, since
+rclone v1.72.0 there is a fallback to normal listing (the same one performed by `rclone size`).
```
Not sure which version I should use here. It's this one correct? Or shall I change it to v1.71.1?
